assertion/function/assertiontree/util.go

```diff
@@ -199,171 +164,144 @@ func AddNilCheck(pass *analysishelper.EnhancedPass, expr ast.Expr) (trueCheck, f
-					), noop, false
+				xLenArgs, yLenArgs := extractLenArgs(x, true /* allowNested */), extractLenArgs(y, true /* allowNested */)
+				if len(xLenArgs) != 0 && len(yLenArgs) != 0 {
+					return produceNegativeNilChecks(slices.Concat(xLenArgs, yLenArgs)...), noop, false
 				}
 				return noop, noop, true
 			},
 		},
-		{ // this exprCheck matches on expressions like `len(a) == 37` or `len(a) == b`
+		{
+			// `len(a) - 1 + b == [positive-int]`
+			// Automatic cases:
+			//   - `[positive-int] == len(a) - 1 + b`
+			//   - `len(a) - 1 + b != [positive-int]`
+			//   - `[positive-int] != len(a) - 1 + b`
 			op: token.EQL,
 			matcher: func(x, y ast.Expr) (RootFunc, RootFunc, bool) {
-				if lenArg, isLen := asLenCall(x); isLen && (isLiteralPositiveInt(y) || isNonLiteralInt(y)) {
-					return produceNegativeNilCheck(lenArg), noop, false
+				if lenArgs := extractLenArgs(x, true /* allowNested */); len(lenArgs) == 1 && likelyPositiveInt(pass, y) {
+					return produceNegativeNilChecks(lenArgs[0]), noop, false
 				}
 				return noop, noop, true
 			},
 		},
-		{ // this exprCheck matches on expressions like `len(a) > 0` or `len(a) > 9`
+		{
+			// `len(a) - 1 + b > [0 or positive-int]`
+			// Automatic cases:
+			//   - `[0 or positive-int] < len(a) - 1 + b`
+			//   - `len(a) - 1 + b <= [0 or positive-int]`
+			//   - `[0 or positive-int] >= len(a) - 1 + b`
 			op: token.GTR,
 			matcher: func(x, y ast.Expr) (RootFunc, RootFunc, bool) {
-				if lenArg, isLen := asLenCall(x); isLen && (isLiteralZeroInt(y) || isLiteralPositiveInt(y) || isNonLiteralInt(y)) {
-					return produceNegativeNilCheck(lenArg), noop, false
+				if lenArgs := extractLenArgs(x, true /* allowNested */); len(lenArgs) == 1 && (pass.IsZero(y) || likelyPositiveInt(pass, y)) {
+					return produceNegativeNilChecks(lenArgs[0]), noop, false
 				}
 				return noop, noop, true
 			},
 		},
-		{ // this exprCheck matches on expressions like `len(a) >= 19`
+		{
+			// `len(a) - 1 + b >= [positive-int]`
+			//
+			// Automatic cases:
+			//   - `[positive-int] <= len(a)`
+			//   - `len(a) < [positive-int]`
+			//   - `[positive-int] > len(a)`
 			op: token.GEQ,
 			matcher: func(x, y ast.Expr) (RootFunc, RootFunc, bool) {
-				if lenArg, isLen := asLenCall(x); isLen && (isLiteralPositiveInt(y) || isNonLiteralInt(y)) {
-					return produceNegativeNilCheck(lenArg), noop, false
+				if lenArgs := extractLenArgs(x, true /* allowNested */); len(lenArgs) == 1 && likelyPositiveInt(pass, y) {
+					return produceNegativeNilChecks(lenArgs[0]), noop, false
 				}
 				return noop, noop, true
 			},
 		},
 	}
 
-	// this applies each of the checkers to see if we can use it to trigger a return from this function
-	// the converse, inverse, and contrapositive of each exprCheck is checked as well
+	// Apply the checkers.
 	for _, check := range checkers {
-		if binExpr.Op == check.op {
+		// Note that `op` might be equal to `converse(op)` (e.g., `==` = `==`), so we must check
```

Comment:
```suggestion
		// Note that `op` might be equal to `converse(op)` (e.g., `==` = `!=`), so we must check
```

I suppose you meant to show that EQ and NEQ are the same?

Reply:
It's actually `==` (op) and `==` (converse of op). Maybe this is just too confusing, moved the example here :)